numpy/_core/src/multiarray/arraywrap.c

```diff
@@ -0,0 +1,308 @@
+/*
+ * Definitions for dealing with array-wrap or array-prepare.
+ */
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#define _MULTIARRAYMODULE
+#define _UMATHMODULE
+
+#include <Python.h>
+
+#include "numpy/arrayobject.h"
+#include "numpy/npy_3kcompat.h"
+#include "get_attr_string.h"
+
+#include "arraywrap.h"
+#include "multiarraymodule.h"
+
+
+/*
+ * Find the array wrap or array prepare method that applies to the inputs.
+ * outputs should NOT be passed, as they are considered individually while
+ * applying the wrapping.
+ *
+ * @param nin number of inputs
+ * @param inputs Original input objects
+ * @param out_wrap Set to the python callable or None (on success).
+ * @param out_wrap_type Set to the type belonging to the wrapper.
+ */
+NPY_NO_EXPORT int
+npy_find_array_wrap(
+        int nin, PyObject *const *inputs,
+        PyObject **out_wrap, PyObject **out_wrap_type)
+{
+    PyObject *wrap = NULL;
+    PyObject *wrap_type = NULL;
+
+    double priority = 0;  /* silence uninitialized warning */
+
+    for (int i = 0; i < nin; i++) {
+        PyObject *obj = inputs[i];
+        if (PyArray_CheckExact(obj)) {
+            if (wrap == NULL || priority < 0) {
```

Comment:
Priority cannot be `< 0` here, right? Should the initialization above be at `-inf`? A benefit of that is that I think the test for `wrap == NULL` would not be needed any more.

Reply:
I'd also write `priority < NPY_PRIORITY` and use that below for `priority =  NPY_PRIORITY` - somewhat easier to read than having to know the priority of arrays is 0.

Reply:
Priority can be `< 0`, this makes sense for example for `memmap` and is used there.  

I had initially thought about it, and decided it seemed dubious.  Maybe all it changes is that `-inf` really means that `__array_wrap__` should *never* be called, which would make sense (and probably nobody uses that anyway).

There is the odd code, that the old code just *skipped* ndarrays, which is worse, because negative wraps take precendence over ndarray!  (yes this is a change here)

Reply:
(Will replace the `NPY_PRIORITY` to not hardcode the 0)

Reply:
But if you initialize `priority = 0`, doesn't that mean that `np.memmap.__array_wrap__` never gets selected? Is that OK?

Reply:
The code here always uses the first one (if there is more than one).  That is what the `NULL` check does.

Reply:
Ah, that is nicer though of course!  I should write `i == 0` that is much clearer...

Reply:
Wait no, arg... that doesn't work, how about just adding a comment for what the `NULL` check does early on?